### PR TITLE
apiextensions: add nullable support to OpenAPI v3 validations

### PR DESCRIPTION
**Pull request by sttts**
OpenAPI v3 supports `nullable`, but not `"type": "null"` (in contrast to JSON Schema). We don't support either today. This PR completes our support to match OpenAPI v3.

**Why we want this:** Empty `RawExtensions` render to null, same for zero-valued timestamps (like in ObjectMeta). I.e. any kind of embedding of Kubernetes types into CRDs leads to potential null values.

Note: we always rejected `"type": "null"` although go-openapi supports that.

```release-note
Add `nullable` support to CustomResourceDefinition OpenAPI validation schemata.
```

Part of the CRD openapi publishing efforts alongside https://github.com/kubernetes/kubernetes/pull/71192 (https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-publish-crd-openapi.md), here to be able to express our basic types like `ObjectMeta` and `RawExtension`.

TODO follow-ups:
- [ ] add CRD tests for `type: object, properties: ....`, `type:[object,null], properties: ....`, `properties: ...` (compare https://github.com/kubernetes/kubernetes/pull/74804#discussion_r261672362) when https://github.com/kubernetes/kubernetes/pull/71192 merges.

### CODE REVIEW

**Review by s-urbaniak — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go [hunk 1]
```diff
@@ -143,5 +143,9 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				c.Fuzz(&obj.Property)
 			}
 		},
+		func(obj *int64, c fuzz.Continue) {
+			// JSON only supports 53 bits because everything is a float
+			*obj = int64(c.Int() & ((1 << 53) - 1))
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go]: not sure why this never flakes, or maybe it did.

**Comment by nikhita** [comment]
I'll take a look tonight.

/label api-review

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 2]
```diff
@@ -641,7 +645,10 @@ func (v *specStandardValidatorV3) validate(schema *apiextensions.JSONSchemaProps
 	}
 
 	if schema.Type == "null" {
-		allErrs = append(allErrs, field.Forbidden(fldPath.Child("type"), "type cannot be set to null"))
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("type"), "type cannot be set to null, use nullable as an alternative"))
+	}
+	if schema.Nullable && schema.Type != "object" && schema.Type != "array" {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: what do we generate for pointers to scalars?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 3]
```diff
@@ -62,6 +62,10 @@ func ConvertJSONSchemaProps(in *apiextensions.JSONSchemaProps, out *spec.Schema)
 	if in.Type != "" {
 		out.Type = spec.StringOrArray([]string{in.Type})
 	}
+	if in.Nullable {
+		// by validation, in.Type is either "object" or "array"
+		out.Type = append(out.Type, "null")
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: I thought `type` could only be a string in openapi v2 and v3 (per https://docs.google.com/spreadsheets/d/1Mkm9L7CXGvRorV0Cr4Vwfu0DH7XRi24YHPiDK1NZWo4/edit#gid=0)

**Comment by liggitt** [comment]
openapi v2 doesn't support nullable, does ConvertJSONSchemaPropsToOpenAPIv2Schema need to drop this field? does dropping it for openapi v2 make the v2 schema weaker or more restrictive?

**Comment by sttts** [comment]
> openapi v2 doesn't support nullable, does ConvertJSONSchemaPropsToOpenAPIv2Schema need to drop this field? does dropping it for openapi v2 make the v2 schema weaker or more restrictive?

Yes, we will have to add it to `ConvertJSONSchemaPropsToOpenAPIv2Schema`. Did that merge already? Will update, if yes.

Nullable implies that we drop the other type as well to make the schema weaker. In v3 we will have it published 1:1.

**Comment by sttts** [comment]
@liggitt updated 
`ConvertJSONSchemaPropsToOpenAPIv2Schema`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: there we usually have omitempty. We could add that as well here, but I have never seen such a use-case. Have you?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: go-openapi is different. It supports JSON Schema like type slices, and it supports null there (but no nullable).

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: I have now generalized nullable to any type (also without where it has no effect, but is perfectly valid). Please take a look.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go [hunk 4]
```diff
@@ -62,12 +62,11 @@ func ConvertJSONSchemaPropsToOpenAPIv2Schema(in *apiextensions.JSONSchemaProps)
 			}
 		}
 
-		if len(p.Type) > 1 {
-			// https://github.com/kubernetes/kube-openapi/pull/143/files#diff-62afddb578e9db18fb32ffb6b7802d92R272
-			// We also set Properties to null to enforce parseArbitrary at https://github.com/kubernetes/kube-openapi/blob/814a8073653e40e0e324205d093770d4e7bb811f/pkg/util/proto/document.go#L247
+		switch {
+		case len(p.Type) == 2 && (p.Type[0] == "null" || p.Type[1] == "null"):
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go]: do we have a test that `type:["object","null"], properties: ...` behaves correctly? (the only difference between this and the previous behavior is that properties is preserved, right?)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 5]
```diff
@@ -70,10 +70,9 @@ func ConvertJSONSchemaPropsWithPostProcess(in *apiextensions.JSONSchemaProps, ou
 	out.Description = in.Description
 	if in.Type != "" {
 		out.Type = spec.StringOrArray([]string{in.Type})
-	}
-	if in.Nullable {
-		// by validation, in.Type is either "object" or "array"
-		out.Type = append(out.Type, "null")
+		if in.Nullable {
+			out.Type = append(out.Type, "null")
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: we only append `null` if type is not ""? does validation enforce type being non-empty?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: or does type:"" imply arbitrary type, including null?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go]: It lets us go here: https://github.com/kubernetes/kube-openapi/blob/master/pkg/util/proto/document.go#L238. So it is handled the same way as `
type:"object", properties: ..` for the non-null case. The null case is always accepted in kube-openapi validation as far as I see, so nullable=true is implied.

I have started a follow-up TODO list in the description. I would like to have e2e tests with CRDs for these cases.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: yes, empty type means anything. This includes the null case. So nullable is not needed.

**Comment by sttts** [comment]
Green.

@liggitt anything left to do here?

**Pull request merged**